azkaban-common/src/test/java/azkaban/metric/MetricManagerTest.java

```diff
@@ -80,13 +80,18 @@ public void managerEmitterMaintenanceTest() {
    */
   @Test
   public void managerEmitterHandlingTest() throws Exception {
+
+    // metrics use System.currentTimeMillis, so that method should be the millis provider
+    final DateTime aboutNow = new DateTime(System.currentTimeMillis());
+
     this.emitter.purgeAllData();
-    final Date from = DateTime.now().minusMinutes(1).toDate();
+
+    final Date from = aboutNow.minusMinutes(1).toDate();
```

Comment:
It should be ok to use the "simple" `DateTime.now()` as long DateTimeUtils is not in some bad state?

Reply:
I tried this – this makes the test fail consistently:

```java
  @Test
  public void managerEmitterHandlingTest() throws Exception {
    // ADDING THESE TWO LINES:
    final Duration FLOW_FINISHED_TIME = Duration.ofMinutes(2);
    DateTimeUtils.setCurrentMillisOffset(-FLOW_FINISHED_TIME.toMillis());

    this.emitter.purgeAllData();
    final Date from = DateTime.now().minusMinutes(1).toDate();
    this.metric.notifyManager();

    this.emitterWrapper.countDownLatch.await(10L, TimeUnit.SECONDS);
    assertEquals(0, this.emitterWrapper.countDownLatch.getCount());

    final Date to = DateTime.now().plusMinutes(1).toDate();
    final List<InMemoryHistoryNode> nodes = this.emitter.getMetrics("FakeMetric", from, to, false);

    assertEquals("Failed to report metric", 1, nodes.size());
    assertEquals("Failed to report metric", nodes.get(0).getValue(), 4);
  }
```

If I add this after the two added lines it works again:

```java
    DateTimeUtils.setCurrentMillisSystem();
```

My suggestion for the eventual fix would be to only modify `ExecutionFlowDaoTest.java` by adding this into its existing `@After` method:

```java
    DateTimeUtils.setCurrentMillisSystem();
```

I found one more leaking test which is `BasicTimeCheckerTest` – it should also reset DateTimeUtils in an `@After` method.

Reply:
`DateTime.now()` takes the milliseconds from the effective joda millis provider [1]. Since metrics use the system millis provider directly [citation needed :D], I suggest the `MetricManagerTest` use the system millis provider as well.

[1] https://github.com/JodaOrg/joda-time/blob/master/src/main/java/org/joda/time/DateTimeUtils.java#L112

Reply:
joda `DateTime` and `minus/plusMinutes` are used, though! :)

Reply:
Yeah it just adds one line of code more. To me there’s no difference if using joda time directly because underneath it should also be calling System#currentTimeMillis. Also it’s safer to get the current time again _after_ the emitter has been called. This shouldn’t really happen, but if the metric call would take more that one minute this could fail because the to date is not after that.